### PR TITLE
Add GraphQL.Dynamic for distributed queries

### PR DESCRIPTION
**Pull request by elauffenburger**
@joemcbride @ravgupms: Here's the PR for our discussion in #582 

**Obviously** this isn't ready for prod (and we've talked about how to make this netstandard compliant), but this should give you an idea of what I was thinking.  I don't _think_ it should be too much work to use Roslyn, so I believe a lot of the core implementation could remain the same if we're in agreement 😄

### CODE REVIEW

**Comment by elauffenburger** [comment]
@joemcbride: can I revive this?  I'd love to get your input!

**Comment by joemcbride** [comment]
@elauffenburger Hey, sorry I haven't been able to dig into this at all yet.  Still on my "at a future date" list.  ☹️

**Comment by PascalSenn** [comment]
Are you creating types for the whole remote schema?

**Comment by elauffenburger** [comment]
@joemcbride all good! Just wanted to keep the conversation alive :)

@PascalSenn yep! But obviously just custom types (as in, types graphql.net isn't aware of by default)

**Comment by PascalSenn** [comment]
@elauffenburger 
Alright. Alright. 

We made some thoughts about this and there are many different ways to stitch a schema together and all with different ups and downs. 

There is the way you chose with generating the schema based on the introspection. 
Ups: 
- it gives the possibility to actually stitch a schema together. That means with some engineering you could also automatically merge types. 

Downs: 
You have to resolve a type twice one in every backend. 

Another way would be to create another ComplexGraphType in which you configure the endpoint, the prefix,  argument and a root query on the endpoint. 
Then on the introspection you just "bubble" up the introspection and decorate it with a prefix. The resolver also just bubbels the resolved json up. 

Downs: 
You'd only be able to stitich together queries and not able to merge trees

Ups: 
the resolving cost would be less cost intensive. 

Conclusion: 
I guess it wouldn't be too bad to have both implementations at the end. Because the use cases are slightly different. What do you think @joemcbride ?

Some follow up questions: 
How hard is it to implement a dataloader in your solution? 
Do you prefix and translate types? 
How about arguments?

**Comment by OneCyrus** [comment]
this looks nice @elauffenburger! as this PR doesn't change anything in graphql-dotnet itself maybe we can polish this as a separate package which doesn't need to live inside the main repo?

we are really interested to have proper schema stitching in graphql-dotnet. this seems to be a good start and with some effort we should be able to move the codegen to roslyn and improve on some edge cases.

File: src/GraphQL.Dynamic.Tests/BasicTests.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.id").Value<object>());
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.company").Value<object>());
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.repos").Value<object>());
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.repos[0].id").Value<object>());
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.repos[0].name").Value<object>());
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.repos[1].id").Value<object>());
+            Assert.NotNull(queryResult.SelectToken("githubDynamic.user.repos[1].name").Value<object>());
+        }
+
+        private string GetFileContentsForTestFile(string relativePath) => File.ReadAllText(Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), relativePath));
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic.Tests/BasicTests.cs]: btw. this gives the wrong path (shadowed assembly path). using the following gives the correct path:

```Path.GetDirectoryName(Uri.UnescapeDataString(new Uri(Assembly.GetExecutingAssembly().CodeBase).AbsolutePath))```

File: src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+            }
+
+            return schema != null;
+        }
+
+        private static Introspection.Schema FetchRemoteSchemaViaHttp(string url)
+        {
+            using (var client = new HttpClient())
+            {
+                var query = new
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs]: just had some issues with mediaheaders. this helps to set the accept format.

```
client.DefaultRequestHeaders
  .Accept
  .Add(new MediaTypeWithQualityHeaderValue("application/json"));
```

File: src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                {
+                    Query = SchemaIntrospection.IntrospectionQuery
+                };
+
+                var response = client.PostAsync(new Uri(url), new StringContent(JsonConvert.SerializeObject(query)))
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs]: as above with the accept header, setting the content type helped to solve an issue in our environment:
```
var response = client.PostAsync(new Uri(url), new StringContent(JsonConvert.SerializeObject(query), Encoding.UTF8, "application/json"))
```

**Comment by elauffenburger** [comment]
Awesome! I'm really glad this is gaining some traction (and thanks to @PascalSenn for such an in-depth analysis 😄)

I actually have a number of changes and bug fixes that I need to merge into this PR (@OneCyrus has already found most of them!)

I'll try to do that ASAP so no one is churning on this too much!

CC @joemcbride

File: src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        private static void PopulateMappings()
+        {
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.String, typeof(string), CreateFieldType<StringGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Int, typeof(int), CreateFieldType<IntGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Long, typeof(long), CreateFieldType<IntGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Float, typeof(float), CreateFieldType<FloatGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Double, typeof(double), CreateFieldType<FloatGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Boolean, typeof(bool), CreateFieldType<BooleanGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Guid, typeof(Guid), CreateFieldType<IdGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.DateTimeOffset, typeof(DateTimeOffset), CreateFieldType<DateGraphType>);
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs]: DateTime is missing:

```
PopulateTypeMapping(LiteralGraphTypeMemberInfoType.DateTime, typeof(bool), CreateFieldType<DateGraphType>);
```

**Comment by OneCyrus** [comment]
that sounds great! btw. did you look into changing to roslyn code generation yet? we are a bit stuck with the current version as we would need this in a .net core application.

and another feature would be nice to have is custom scalar types.

File: src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs (hunk 4 above)
Comment (PascalSenn) [reply src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs]: Also: 
I'm not so sure about this line 
```csharp
PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Guid, typeof(Guid), CreateFieldType<IdGraphType>);
```
because a GUID is not certainly a `IdGraphType` as a `IdGraphType` is not always a GUID
https://github.com/graphql-dotnet/graphql-dotnet/blob/c4bc7ca006d3ab5d7e60089c12e8e7038ee3797f/src/GraphQL/Types/IdGraphType.cs#L27

**Comment by OneCyrus** [comment]
two more questions about using this library:

- is there an easy way to "generate" the query which i would send to the remote server inside the RemoteField resolver? something like generate query from resolver "ctx.FieldAST" (with children) and map this to the typeName from the remote schema.

- did you already experiment with filtering variables/arguments to only pass the correct ones to the remote server?

**Comment by PascalSenn** [comment]
@OneCyrus  

- You can use the AstPrinter to Print a graphql query `AstPrinter.Print(context.FieldAst.SelectionSet)`

**Comment by OneCyrus** [comment]
@PascalSenn that's nice! now we need to have an AstPrinter which can print the AST in the other direction in an other schema.

**public schema**
```
Query {
   LocalType {
      RemoteType {
         MyRemoteField
      }
      OtherRemoteType {
         OtherRemoteField
      }
   }
}
```

**schema from remote server**
```
Query {
   RemotePrimaryType {
     EvenDeeper {
        RemoteType {
           MyRemoteField
        }
        NotQueriedRemoteType {
           NotQueriedField
        }
     }
   }
}
```

**AST printer would give me**
```javascript
{
   RemoteType {
       MyRemoteField
    }
}
```

**and this is what i would need**
``` 
Query {
  RemotePrimaryType {
    EvenDeeper {
      RemoteType {
        MyRemoteField
      }
    }
  }
}
```

though it could get difficult if there are multiple times the same type in the remote schema. GraphlQL.Dynamic currently has a type as en entry point. And in this direction it's not necessary to know where this type is in the AST. but resolving it back to the remote schema we are missing the path to the type. as in the test case in this PR (https://github.com/graphql-dotnet/graphql-dotnet/pull/592/files#diff-3c2a7abce1b810eb64bdad8bfd0ea9b5R21) "GithubAPI"/GithubMoniker should probably be a path and not just a type/name.

**Comment by PascalSenn** [comment]
@OneCyrus  
Funny enough. I tried to get all variable references to pass down variables for stitching and was going to refactor ASTPrinter. Would have been a good moment to add a reverse printer.
But @joemcbride  showed me the BasicVisitor. 

I guess we could use something like this. 
https://github.com/graphql-dotnet/graphql-dotnet/issues/690 

Just pattern match around the different node types should do the job

**Comment by joemcbride** [comment]
May want to have a look at how `DocumentValidator` walks the tree.  It uses `TypeInfo` to keep track of context within the document tree (so you can look at parent types and such).

https://github.com/graphql-dotnet/graphql-dotnet/blob/3f2355aacd6cfd8e080dba1d802b2fa00f958f85/src/GraphQL/Validation/DocumentValidator.cs#L35-L59

**Comment by OneCyrus** [comment]
@elauffenburger in the next couple of days we would like to start with adopting this approach for our project. would it be possible for you as mentioned to address the known bugs or should we fork this from the current state?

@PascalSenn sounds good!

**Comment by OneCyrus** [comment]
@elauffenburger I just created a PR to change to roslyn code generation and .net standard

https://github.com/elauffenburger/graphql-dotnet/pull/1

File: src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+                // get type
+                if (!TryGetFieldTypeFromRemoteSchema(schema, _name, out var type))
+                {
+                    // if no type found: fail
+                    // TODO: fail better
+                    throw new Exception($"Failed to find type '{_name}' in remote '{_remoteLocation}' schema");
+                }
+
+                Name = $"{_remoteLocation}.{_name}";
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs]: this generates not valid names. changing to ```Name = $"{_name}";``` should fix this. or is there a reason other than duplicated types for this?

File: src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+                if (!RemoteServerTypes.TryGetValue(remote, out var remoteTypes))
+                {
+                    return null;
+                }
+
+                var schemaType = remoteTypes.FirstOrDefault(t => t.Name == literalMember.TypeName);
+                var realType = literalMember.IsList
+                    ? typeof(ListGraphType<>).MakeGenericType(schemaType)
+                    : schemaType;
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs]: this fails if the schemaType is a base GraphQL type e.g. a ListGraphType<StringGraphType>

**Comment by elauffenburger** [comment]
Hey guys, sorry I dropped off the face of the Earth; a bit busy at the day job :)

I'd like to address the bugs here to unblock @onecyrus & co, so that'll be my first priority; I'll port everything over from our codebase to the PR and then address comments brought up here.

Thanks for everyone's support!

cc: @joemcbride

File: src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs (hunk 4 above)
Comment (elauffenburger) [reply src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs]: @OneCyrus: fixed as part of my merges; thanks for the great find 😄 
@PascalSenn: yeah, honestly I don't know the GraphQL spec well enough to know the right way to handle this.  I just know that it's supposed to be a unique value that serializes as a string -- any help would be appreciated here!

File: src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs (hunk 5 above)
Comment (elauffenburger) [reply src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs]: yep; you've got it -- it's to avoid name collisions.  This should be fixed as part of merging my changes in!

File: src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs (hunk 6 above)
Comment (elauffenburger) [reply src/GraphQL.Dynamic/LiteralGraphType/RemoteLiteralGraphType.cs]: Again, great catch; this is fixed per my incoming fix commits

**Comment by elauffenburger** [comment]
Alright, I just kinda sat down and finally did it 😄
Let's give it another look and try to wrap this in the next week or so!

cc: @PascalSenn @OneCyrus @joemcbride

**Comment by OneCyrus** [comment]
@elauffenburger thanks a lot! I'll have a look. 
for the codegen PR I've some additional changes which I'll push there in the next days.

File: src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Double, typeof(double), CreateFieldType<FloatGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Boolean, typeof(bool), CreateFieldType<BooleanGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Guid, typeof(Guid), CreateFieldType<IdGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.DateTime, typeof(DateTimeOffset), CreateFieldType<DateGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.DateTimeOffset, typeof(DateTimeOffset), CreateFieldType<DateGraphType>);
+        }
+
+        internal static Type GetPrimitiveGraphType(string name) => _graphTypesLookup[name]?.GetType();
+
+        internal static string GenerateRemoteTypeName(string remoteLocation, string name) => SanitizeTypeNameRegex.Replace($"{remoteLocation}.{name}", "_");
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs]: putting the full url in front of every remote type gives a very verbose schema. the docs look very busy as well. can't we find a simpler way to solve this? e.g. using the moniker or an intelligent solution which only triggers for conflicting types.

**Comment by OneCyrus** [comment]
i just merged your branch in our internal exploration codebase and played around with it. while it works good for what it does right now we still have three major limitations which we need to tackle before we can use it for real use cases.

- Arguments: currently the inputtypes are not assigned to the graphtypes. do you have any ideas how to do this the best way?

- UNION: we are using unions and would need to have some basic union support. from your comments ii get that you made some thoughts about them but they are not that easy to implement. we thought about them a bit as well and probably the only way to do this would be to bet on forcing __typename for unions in the query to the remote location. dynamic deserialization with type lookup should be possible.

- dynamic query builder: fully generate the query from all the information inside the graphql AST and remote schemas to request the data from remote location. did you implement something like that yet or are you using more or less fixed queries to request the data?

i guess we already made quite some progress with this PR and once we have the remaining limitations fixed it will be an awesome solution.

File: src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        private static void PopulateMappings()
+        {
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.String, typeof(string), CreateFieldType<StringGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Int, typeof(int), CreateFieldType<IntGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Long, typeof(long), CreateFieldType<IntGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Float, typeof(float), CreateFieldType<FloatGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Double, typeof(double), CreateFieldType<FloatGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Boolean, typeof(bool), CreateFieldType<BooleanGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.Guid, typeof(Guid), CreateFieldType<IdGraphType>);
+            PopulateTypeMapping(LiteralGraphTypeMemberInfoType.DateTime, typeof(DateTimeOffset), CreateFieldType<DateGraphType>);
```
Comment (OneCyrus) [inline src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphTypeHelpers.cs]: not sure if this should be DateTimeOffset or DateTime

**Comment by elauffenburger** [comment]
@OneCyrus: thanks for taking a look and for giving it a semi-⭐️ of approval 😄 

So to answer your questions:
  - Arguments:
    - I'll need to take a look at this -- I thought I implemented that, but I may have missed it...
  - Unions:
    - Yeah, I haven't really event given it a thought tbh; obviously we'll have to support it, so my best approach (off the top of my head) would be exactly what you suggested.  I don't think there's an issue with deserialization (since we really have *no* idea what the remote is returning us); i think the only challenges will come from supporting them from an "infrastructure" perspective (as in, we have to float the fact that these are unions along to field resolvers).  I don't think it will be too challenging to implemented
  - Dynamic query builder:
    - I don't have a good solution for this yet 😭.  We've just been taking the selection set from the `FieldAst` and throwing that at the remote.

**Comment by PascalSenn** [comment]
@elauffenburger  

- Dynamic query builder:
Thats basically what we are doing too.
I just added #710 yesterday to support inline fragments in AstPrinter.
With `BasicVisitor` it's easy to walk down the tree and get all variable references used in the selection set. 
Based on that we can get all variable definitions that map with the variable references and add them to a new document. We can also add the operation name, type and the selection set. 

To add the `__typename`'s I've got a bit of a controversial idea. 
The problem we face is, that we should not mutate the SelectionSet and I doubt it's easy to clone it. 
We could just make this protected:
https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/AstPrinter.cs#L155

Create our own Printer that extends AstPrinter that automatically includes all typenames: 
https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/AstPrinter.cs#L227-L234

https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/AstPrinter.cs#L393-L399

If we request __typenames in the frontend and request them from the backend and then further from a remote schema, we would request them twice. But I guess that wouldn't really be a problem. If i request:
image

I get the response type only once. 
image

I'm not sure if they acutally get resolved multiple times. But even if, there's little overhead and we could still check if the first item of a selection set is already a __typename. (So in deeply nested schemas we would request them only twice) 

Maybe @joemcbride  has a better idea. 

Maybe interesting for https://github.com/graphql-dotnet/graphql-client/issues/42 @deinok

**Comment by thivy** [comment]
@joemcbride is this part of the main package now?

**Comment by joemcbride** [comment]
@thivy there are no plans to merge this at this time.  This is based on my ability to support this code in the future.  Right now that is not possible.

**Comment by SimonCropp** [comment]
I might be missing something. But why cant this ship as a stand alone nuget and live in its own repo?

**Comment by deinok** [comment]
@elauffenburger Can you give me a TLDR for this, as @PascalSenn said that It could be interesting for the GraphQL Client?

**Comment by OneCyrus** [comment]
@deinok what @PascalSenn meant is that with introspection and code generation (what this PR here actually does), you could do a fully typed graphql client.

@PascalSenn and I are moving over to a competing solution for schema stitching in .Net. We won't invest more into this PR as well.

**Comment by deinok** [comment]
Sure, code generation using something like this : Microsoft.CodeAnalysis.CSharp from the Introspection model as an MSBuild Task would be awesome

File: src/GraphQL.Dynamic/Introspection/Types.cs [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    public static class Serialize
+    {
+        public static string ToJson(this Root self) => JsonConvert.SerializeObject(self, GraphQL.Dynamic.Types.Introspection.Converter.Settings);
+    }
+
+    internal class Converter : JsonConverter
+    {
+        public override bool CanConvert(Type t) => t == typeof(TypeElementTypeKind) || t == typeof(TypeElementTypeKind) || t == typeof(TypeElementTypeKind) || t == typeof(TypeElementTypeKind) || t == typeof(TypeElementTypeKind?) || t == typeof(TypeElementTypeKind?) || t == typeof(TypeElementTypeKind?) || t == typeof(TypeElementTypeKind?);
```
Comment (kalexii) [inline src/GraphQL.Dynamic/Introspection/Types.cs]: Hey, guys. Just browsing through the code, anyone seeing anything weird here?

File: src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphType.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+            if (listType == null)
+            {
+                return null;
+            }
+
+            return listType.GetGenericArguments().FirstOrDefault();
+
+        }
+        private static bool IsListType(Type type) => type.GetInterfaces().Any(i => i == typeof(IEnumerable));
```
Comment (kalexii) [inline src/GraphQL.Dynamic/LiteralGraphType/LiteralGraphType.cs]: string counts too?

**Comment by codestellar** [comment]
This PR is open since long! Will there be any action on it?

**Comment by joemcbride** [comment]
@codestellar no, this won’t be pulled in.  Schema stiching has been superseded by Apollo Federation. See #1169
